service/statusroutes.go

```diff
@@ -25,47 +26,87 @@ import (
+
 // StatusResponse type.
 type StatusResponse struct {
 	Status  string `json:"status"`
 	Name    string `json:"name"`
 	Version string `json:"version"`
 }
 
-func handleStatusRoutes(w http.ResponseWriter, serviceName, serviceVersion string) (*StatusResponse, []byte) {
+func sendStatusRoutes(w http.ResponseWriter, logger *logrus.Entry, ok bool, serviceName, serviceVersion string) {
+	statusMessage := "OK"
+	statusCode := http.StatusOK
+	if !ok {
+		statusMessage = "KO"
+		statusCode = http.StatusServiceUnavailable
+	}
+
 	w.Header().Add(utils.ContentTypeHeaderKey, utils.JSONContentTypeHeader)
 	status := StatusResponse{
-		Status:  "OK",
+		Status:  statusMessage,
 		Name:    serviceName,
 		Version: serviceVersion,
 	}
 	body, err := json.Marshal(&status)
 	if err != nil {
 		w.WriteHeader(http.StatusServiceUnavailable)
-		return nil, nil
+		return
 	}
 
-	return &status, body
+	w.WriteHeader(statusCode)
+	if _, err := w.Write(body); err != nil {
+		logger.WithField("error", logrus.Fields{"message": err.Error()}).Warn("failed response write")
+	}
 }
 
-var statusRoutes = []string{"/-/rbac-healthz", "/-/rbac-ready", "/-/rbac-check-up"}
+func handleOKHandler(serviceName, serviceVersion string) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, req *http.Request) {
+		sendStatusRoutes(
+			w,
+			glogrus.FromContext(req.Context()),
+			true,
+			serviceName,
+			serviceVersion,
+		)
+	}
+}
+
+func handleSDKReadyHandler(sdkBoot *SDKBootState, serviceName, serviceVersion string) func(http.ResponseWriter, *http.Request) {
+	readyFlagLock := &sync.Mutex{}
+
+	sdkReady := false
+	go func(sdkBoot *SDKBootState, readyFlagLock *sync.Mutex) {
+		sdkBoot.Get()
+
+		readyFlagLock.Lock()
+		defer readyFlagLock.Unlock()
+		sdkReady = true
+	}(sdkBoot, readyFlagLock)
 
-func handleStatusEndpoint(serviceName, serviceVersion string) func(http.ResponseWriter, *http.Request) {
 	return func(w http.ResponseWriter, req *http.Request) {
-		_, body := handleStatusRoutes(w, serviceName, serviceVersion)
-		if _, err := w.Write(body); err != nil {
-			logger := glogrus.FromContext(req.Context())
-			logger.WithField("error", logrus.Fields{"message": err.Error()}).Warn("failed response write")
-		}
+		readyFlagLock.Lock()
+		defer readyFlagLock.Unlock()
+
+		sendStatusRoutes(
+			w,
+			glogrus.FromContext(req.Context()),
+			sdkReady,
+			serviceName,
+			serviceVersion,
+		)
 	}
 }
 
 // StatusRoutes add status routes to router.
-func StatusRoutes(r *mux.Router, serviceName, serviceVersion string) {
-	statusEndpointHandler := handleStatusEndpoint(serviceName, serviceVersion)
-	r.HandleFunc("/-/rbac-healthz", statusEndpointHandler)
+func StatusRoutes(r *mux.Router, sdkBoot *SDKBootState, serviceName, serviceVersion string) {
+	sdkReadyHandler := handleSDKReadyHandler(sdkBoot, serviceName, serviceVersion)
+	alwaysOKHandler := handleOKHandler(serviceName, serviceVersion)
+
+	r.HandleFunc("/-/rbac-healthz", alwaysOKHandler)
```

Comment:
If the initialization fails, the health route is ok. Is it correct? The service would be shutdown exiting the process, but it seems not too correct.

Reply:
yes, the service is alive since the web server is up, if something wrong occurs the service gets shut down and so the probe will start to fail; we could link this to the startup state but I don't see any particular added value a part from the fact that we would let someone else kill the process rather than the process itself.

I think its better for the process to stop by itself whenever it can 🤷

Reply:
If we can terminate the process without any side effects from within, I'd prefer to go that way for the sake of code clarity.

Reply:
so you agree with the liveness to be always OK? 🤔

Reply:
Yes, I'd prefer to respond also ko (since the service is not healthy, so it could be more correct) but it's not an issue since the service will kill itself.